core/market/src/db/model/agreement_events.rs

```diff
@@ -0,0 +1,97 @@
+use chrono::{DateTime, NaiveDateTime, Utc};
+use diesel::sql_types::Text;
+use std::fmt::Debug;
+
+use crate::db::model::{AgreementId, OwnerType};
+use crate::db::schema::market_agreement_event;
+
+use ya_client::model::market::agreement_event::AgreementTerminator;
+use ya_client::model::market::{AgreementOperationEvent as ClientEvent, JsonReason};
+use ya_diesel_utils::DatabaseTextField;
+
+#[derive(
+    DatabaseTextField,
+    strum_macros::EnumString,
+    derive_more::Display,
+    AsExpression,
+    FromSqlRow,
+    PartialEq,
+    Debug,
+    Clone,
+    Copy,
+)]
+#[sql_type = "Text"]
+pub enum AgreementEventType {
+    Approved,
+    Rejected,
+    Cancelled,
+    Terminated,
+}
+
+#[derive(Clone, Debug, Queryable)]
+pub struct AgreementEvent {
+    pub id: i32,
+    pub agreement_id: AgreementId,
+    pub event_type: AgreementEventType,
+    pub timestamp: NaiveDateTime,
+    pub issuer: OwnerType,
+    pub reason: Option<String>,
+    pub signature: Option<String>,
+}
+
+#[derive(Clone, Debug, Insertable)]
+#[table_name = "market_agreement_event"]
+pub struct NewAgreementEvent {
+    pub agreement_id: AgreementId,
+    pub event_type: AgreementEventType,
+    pub issuer: OwnerType,
+    pub reason: Option<String>,
+}
+
+impl AgreementEvent {
+    pub fn into_client(self) -> ClientEvent {
+        let agreement_id = self.agreement_id.into_client();
+        let event_date = DateTime::<Utc>::from_utc(self.timestamp, Utc);
+        let reason = self
+            .reason
+            .map(|reason| serde_json::from_str::<JsonReason>(&reason))
+            .map(|result| result.map_err(|e| {
+                log::warn!(
+                    "Agreement Event with not parsable Reason in database. Error: {}. Shouldn't happen \
+                     because market is responsible for rejecting invalid Reasons.", e
+                )
```

Comment:
minor: I'd also return this warning mesage as a proper JsonReason for the clients to be aware and can report if it happens one day

Reply:
You mean, that you would set this error message in JsonResponse?

Reply:
yes, and extend this message to contain whole reason contents